Ducnguyen/bitgobase interface

ducnguyen-bitgo opened the pull request:
1. Defined a BitGoBase interface in `sdk-core` used by baseCoin, enterprises, keychains, markets, pendingApprovals, wallets, and etc.
2. Moved api interface from `sdk-api` to `sdk-core`.
3. Extract keychain, webhooks, wallet, wallets, enterprise, enterprises, pendingApproval, pendingApprovals, and tradingAccount interfaces into `sdk-core`.
4. Added baseCoin interface into `sdk-core`.
5. Updated all coins needed imports to new locations in `sdk-core`.
6. Changed baseCoin and its dependent classes in `bitgo` to implement corresponding interfaces in `sdk-core`.
7. Updated unit test imports.
8. Resolved naming collisions for the types `BlsKeyPair` and `TransactionExplanation` which were defined in both `account-lib` and `bitgo`.
9. `TransactionExplanation` previously defined in `bitgo` is now `ITransactionExplanation`. `TransactionExplanation` previously defined in account-lib extends this generic type.
10. `BlsKeyPair` previously defined in `bitgo` is now `IBlsKeyPair`. `BlsKeyPair` previously defined in `account-lib` remains unchanged.

ducnguyen-bitgo committed:
feat(sdk-core): Define new BitGoBase interface in sdk-core

Defined a new BitGoBase interface in replace of BitGo object.

Ticket: BG-48170

bitgoAaron commented on modules/sdk-core/src/bitgo/bitgoBase.ts:
```diff
@@ -0,0 +1,43 @@
+import * as superagent from 'superagent';
+import { EnvironmentName } from './environments';
+
+export interface BitGoRequest<ResultType = any> extends superagent.SuperAgentRequest {
```
most of these interfaces already exist in `@bitgo/sdk-api`. Those should be moved here, not duplicated.

ducnguyen-bitgo replied:
Do you mean we should create a folder called `sdk-api` in `sdk-core` and create a file i.e `api.ts` to define the BitGoRequest interface?

bitgoAaron replied:
yes

bitgoAaron commented on modules/sdk-core/src/bitgo/bitgoBase.ts:
```diff
@@ -0,0 +1,43 @@
+  toString(): string;
+}
+
+export interface BitGoBase {
```
`BitGoAPI` in `@bitgo/sdk-api` needs to implement this interface.

ducnguyen-bitgo replied:
Should it be part of ticket BG-48170 or should we use a different ticket for this?

bitgoAaron replied:
it should be part of this

bitgoAaron commented on modules/sdk-core/src/bitgo/bitgoBase.ts:
```diff
@@ -0,0 +1,43 @@
+}
+
+export interface BitGoBase {
+  coin(coinName: string): any; // need to change it to BaseCoin once it's moved to @bitgo/sdk-core
```
we shouldn't use `any` here. We need to define the `BaseCoin` interface, which requires defining the other interfaces. 🐔  <-> 🥚

ducnguyen-bitgo replied:
I agree with you on that. The reason I used `any` there was to resolve an issue in one of the classes in `bitgo` module. Also, `BaseCoin` is still in `bitgo` module so I did not want to import `BaseCoin` from `bitgo` since it would create a circular dependency between `bitgo` and `sdk-core`.

bitgoAaron committed:
chore(sdk-core): add placeholder interfaces

- empty interfaces for stubbing out full interfaces later

Ticket: BG-48171

ducnguyen-bitgo committed:
chore(sdk-core): move api interfaces to sdk-core

Moved api interfaces to sdk-core.

Ticket: BG-48170

bitgoAaron committed:
chore(sdk-core): rename files with i prefix

- interface files renamed to use I<classname> convention

Ticket: BG-48171

bitgoAaron committed:
chore(sdk-core): rename classes to i prefix

- interface class names renamed

Ticket: BG-48171

bitgoAaron committed:
chore(sdk-core): extract keychain interface

- adds interface for iKeychains

Ticket: BG-48410

bitgoAaron committed:
chore(sdk-core): extract webhooks interface

- adds interface for iWebhooks

Ticket: BG-48407

bitgoAaron committed:
chore(sdk-core): add iBaseCoin

- add IBaseCoin interface
- add iTssUtils & other required mpc interfaces

Ticket: BG-48412

ducnguyen-bitgo committed:
chore(sdk-core): extract wallet interface

- Added interface for wallet.

Ticket: BG-48404

ducnguyen-bitgo committed:
chore(sdk-core): extract wallets interface

- Added interface for wallets.

Ticket: BG-48404

ducnguyen-bitgo committed:
chore(sdk-core): extract enterprise and enterprises interfaces

- Added interfaces for enterprise and enterprises

Ticket: BG-48405

ducnguyen-bitgo committed:
chore(sdk-core): extract PendingApproval and PendingApprovals interface

- Added interfaces for pending approval and pending approvals.

Ticket: BG-48406

bitgoAaron committed:
chore(sdk-core): extract TradingAccount interface
- interface files for TradingAccount
Ticket: BG-48171

bitgoAaron committed:
chore(sdk-core): move interfaces into folders

- interfaces & implementations will be grouped in same folder

Ticket: BG-48171

bitgoAaron committed:
chore(bitgo): implement all interfaces from core

- baseCoin and all dependency classes implement respective interfaces
- updates necessary imports & missing types
- adds default recoverToken metho to baseCoin to work around ETH dep

Ticket: BG-48171

bitgoAaron committed:
chore(express): update import

- import CustomSigningFunction from core

Ticket: BG-48171

bitgoAaron committed:
chore(sdk-core) reconcile duplicate types

- TransactionExplanation existed in both account-lib & core
- duplicate replaced with concrete implementation of the generic type
- fixes type imports in express clientRoutes

Ticket: BG-48171

bitgoAaron committed:
chore(bitgo): update coin imports

- all the coins needed import updates to new locations in core

Ticket: BG-48171

bitgoAaron committed:
chore(bitgo): update unit test imports

- fixes broken imports in bitgo unit tests

Ticket: BG-48171

bitgoAaron committed:
chore(express): re-export types from bitgo

- express requires installing from the public version of bitgo
- until that includes these sdk-core changes, we have to
  make backwards compatible exports

Ticket: BG-48171

bitgoAaron committed:
chore(sdk-core): add openpgp

- updates sdk-core dependencies

Ticket: BG-48171
